### PR TITLE
Regression testing: Token type, replay, and lifetime validation.

### PR DESCRIPTION
**Pull request by iNinja**
# Regression testing: Token type, replay, and lifetime validation.

- Updated theory, added new test cases.
 
Part of #2711

### CODE REVIEW

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs [hunk 1]
```diff
@@ -354,22 +479,48 @@ static Dictionary<string, object> AdditionalEcdhEsHeaderParameters(JsonWebKey pu
             }
         }
 
+        private class TestTokenReplayCache : ITokenReplayCache
+        {
+            private bool _onAddReturnValue;
+            private bool _onFindReturnValue;
+            public TestTokenReplayCache(bool onAddReturnValue, bool onFindReturnValue)
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs:486]:
```suggestion

            public TestTokenReplayCache(bool onAddReturnValue, bool onFindReturnValue)
```

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs [hunk 2]
```diff
@@ -354,22 +479,48 @@ static Dictionary<string, object> AdditionalEcdhEsHeaderParameters(JsonWebKey pu
...
+            {
+                _onAddReturnValue = onAddReturnValue;
+                _onFindReturnValue = onFindReturnValue;
+            }
+            public bool TryAdd(string securityToken, DateTime expirationTime)
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs:491]:
```suggestion

            public bool TryAdd(string securityToken, DateTime expirationTime)
```

**Review by keegan-caruso — APPROVED** [review]
Few spacing issues, otherwise LGTM

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs [hunk 3]
```diff
@@ -13,6 +13,8 @@
 using Microsoft.IdentityModel.Tokens;
 using Xunit;
 
+#nullable enable
+
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs:17]: be consistent with spacing around #nullable.
Other files do not have a blank line.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs [hunk 4]
```diff
@@ -102,6 +109,11 @@ public static TheoryData<JsonWebTokenHandlerValidationParametersTheoryData> Json
                         ValidationParameters = CreateValidationParameters(
                             Default.Issuer, [Default.Audience], Default.AsymmetricSigningKey),
                     },
+                    new JsonWebTokenHandlerValidationParametersTheoryData("Invalid_ValidationParametersNull")
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs:112]: add a new line between creation of theoryData.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs [hunk 5]
```diff
@@ -23,6 +25,7 @@ public async Task ValidateTokenAsync(JsonWebTokenHandlerValidationParametersTheo
             var context = TestUtilities.WriteHeader($"{this}.ValidateTokenAsync", theoryData);
 
             JsonWebTokenHandler jsonWebTokenHandler = new JsonWebTokenHandler();
+            jsonWebTokenHandler.SetDefaultTimesOnTokenCreation = theoryData.SetDefaultValuesOnTokenCreation;
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerValidationParametersTests.cs:28]: Why do we not create the jwt when creating the theoryData?

**Comment by brentschmaltz** [comment]
If the goal of this test is to check that the path with TokenValidationParameters is the same as ValidationParameters, the number of tests will be large.

I would suggest having separate files for each validation step.
In this way we will be able to focus on one validation step at time.
This separation will result in fewer merge conflicts when multiple engineers will work on validation.

**Review by brentschmaltz — CHANGES_REQUESTED** [review]
Please split into PR's focusing on one validation step.

**Comment by iNinja** [comment]
Closing the PR, will re-raise individual ones as I add more tests.

**Pull request closed without merging**